# Analyzing Used Car Listings on eBay Kleinanzeigen

I'll will be working on a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The version of the dataset I'll be working with is a sample of 50,000 data points.
The data dictionary provided with data is as follows:

   * dateCrawled - When this ad was first crawled. All field-values are taken from this date.
   * name - Name of the car.
   * seller - Whether the seller is private or a dealer.
   * offerType - The type of listing
   * price - The price on the ad to sell the car.
   * abtest - Whether the listing is included in an A/B test.
   * vehicleType - The vehicle Type.
   * yearOfRegistration - The year in which which year the car was first registered.
   * gearbox - The transmission type.
   * powerPS - The power of the car in PS.
   * model - The car model name.
   * odometer - How many kilometers the car has driven.
   * monthOfRegistration - The month in which which year the car was first registered.
   * fuelType - What type of fuel the car uses.
   * brand - The brand of the car.
   * notRepairedDamage - If the car has a damage which is not yet repaired.
   * dateCreated - The date on which the eBay listing was created.
   * nrOfPictures - The number of pictures in the ad.
   * postalCode - The postal code for the location of the vehicle.
   * lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

Therefore, I'll start by cleaning the column names to make the data easier to work with.

In [2]:
autos.columns


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

I'll make a few changes here:

   * Change the columns from camelcase to snakecase.
   * Change a few wordings to more accurately describe the columns.



In [3]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

The initial observations:

   There are a number of text columns where all (or nearly all) of the values are the same:
   * seller
   * offer_type
   * The num_photos column looks odd, so I'll check this out


In [5]:
autos['num_pictures'].value_counts()

0    50000
Name: num_pictures, dtype: int64

The num_pictures column has 0 for every row, so I'll drop this column with the seller and offer_type columns

In [6]:
autos = autos.drop(['num_pictures','seller','offer_type'], axis=1)

Also, there are two columns, price and auto, which are numeric values with extra characters being stored as text. I'll clean and convert these to numeric dtype.

In [7]:
autos['price'] = (autos['price']
                  .str.replace('$','')
                  .str.replace(',','')
                  .astype(int)
                 )

autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [8]:
autos['odometer'] = (autos['odometer']
                     .str.replace('km','')
                     .str.replace(',','')
                     .astype(int)
                    )

autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [9]:
#Renaming the odometer column
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Exploring the odometer and price columns

In [10]:
print(autos['odometer_km'].unique().shape)

autos['odometer_km'].describe()

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

The above result shows that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [11]:
print(autos['price'].unique().shape)

print(autos['price'].value_counts().head(10))

autos['price'].describe()

(2357,)
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, I may consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, I have to check out the highest prices further.


In [12]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [13]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. I will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.


In [14]:
autos = autos[autos['price'].between(1,351000)]

autos.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

## Exploring the date columns

There are a number of columns with date information:

   * date_crawled
   * registration_month
   * registration_year
   * ad_created
   * last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

In [15]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [16]:
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform

In [17]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which can be used to determine on what day a listing was removed, presumably because the car was sold.

From the above results, the last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.


In [18]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [19]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [20]:
print(autos['registration_year'].unique().shape)

print(autos['registration_year'].value_counts().sort_index())

(95,)
1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64


The year that the car was first registered will likely indicate the age of the car. Looking at this column, I noted some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

## Dealing with Incorrect Registration Year Data

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s

In [21]:
(~autos["registration_year"]
 .between(1900,2016)
).sum() / autos.shape[0]

0.038793369710697

In order to make a decision to remove the listings with the values below 1900 and above 2016, I have to determine what percentage of our data has invalid values in this column using the code above. The returned result showed that this is less than 4% of the data, therefore I will remove these rows.

In [22]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring Price by Brand

In [23]:
print(autos['brand'].unique().shape)
print('\n')
print(autos['brand'].value_counts())
print('\n')
print(autos['brand'].value_counts(normalize=True))


(40,)


volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64


volkswagen        0.211264
bmw              

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so I will limit the analysis to brands representing more than 5% of total listings

In [24]:
brand_count = autos['brand'].value_counts(normalize=True)
high_brands = brand_count[brand_count>0.05].index
print(high_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [25]:
brand_mean_prices = {}

for b in high_brands:
    select_rows = autos[autos['brand'] == b]
    mean_price = select_rows['price'].mean()
    brand_mean_prices[b] = int(mean_price)
print(brand_mean_prices)

{'bmw': 8332, 'mercedes_benz': 8628, 'volkswagen': 5402, 'ford': 3749, 'audi': 9336, 'opel': 2975}


Of the top 5 brands, there is a distinct price gap:

   * Audi, BMW and Mercedes Benz are more expensive
   * Ford and Opel are less expensive
   * Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.





## Exploring Mileage

In [26]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending=False)
pd.DataFrame(bmp_series, columns = ['mean_price'])


mean_price
audi                 9336
mercedes_benz        8628
bmw                  8332
volkswagen           5402
ford                 3749
opel                 2975

In [27]:
brand_mean_mileage = {}

for b in high_brands:
    select_row = autos[autos['brand'] == b]
    mean_mileage = select_row['odometer_km'].mean()
    brand_mean_mileage[b] = int(mean_mileage)
print(brand_mean_mileage)
   
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending=False)
pd.DataFrame(bmm_series, columns = ['mean_mileage'])

{'bmw': 132572, 'mercedes_benz': 130788, 'volkswagen': 128707, 'ford': 124266, 'audi': 129157, 'opel': 129310}


mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [28]:
brand_info = pd.DataFrame(bmp_series, columns = ['mean_price'])
brand_info['mean_mileage'] = bmm_series
brand_info

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

## Exploring the Gearbox, Fuel_type and Unrepaired_damage columns

The gearbox,fuel_type and unrepaired_damage columns have data that uses german words, therefore I used the Series.map method to map the values and translate them to their English counterparts

In [29]:
gear_map = {
    'manuell':'manually', 
    'automatik':'automatic', 
    'NaN':'not available'
}

autos['gearbox'] = autos['gearbox'].map(gear_map)


In [30]:
autos['gearbox'].value_counts(dropna=False)

manually     34715
automatic     9856
NaN           2110
Name: gearbox, dtype: int64

In [31]:
fuel_type_map = {
    'benzin':'petrol',
    'diesel':'diesel',
    'NaN':'not indicated',
    'lpg':'lpg',
    'cng':'cng',
    'hybrid':'hybrid',
    'elektro':'electro',
    'andere':'other'
}

autos['fuel_type'] = autos['fuel_type'].map(fuel_type_map)

In [32]:
autos['fuel_type'].value_counts(dropna=False)

petrol     28540
diesel     14032
NaN         3318
lpg          649
cng           71
hybrid        37
electro       19
other         15
Name: fuel_type, dtype: int64

In [33]:
unrepaired_damage_map ={'nein':'no','ja':'yes'}

autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage_map,na_action='ignore')

In [34]:
autos['unrepaired_damage'].value_counts(dropna=False)

no     33834
NaN     8307
yes     4540
Name: unrepaired_damage, dtype: int64

## Coverting the date column from string object to numeric data

In [35]:
#converting the date_crawled column to numeric data
autos['date_crawled'] = (autos['date_crawled'].str[:10]
 .str.replace('-','')
 .str.replace('-','')
 .astype(int)
)

#converting the ad_created column to numeric data
autos['ad_created'] = (autos['ad_created'].str[:10]
 .str.replace('-','')
 .str.replace('-','')
 .astype(int)
)

#converting the last_seen column to numeric data
autos['last_seen'] = (autos['last_seen'].str[:10]
 .str.replace('-','')
 .str.replace('-','')
 .astype(int)
)


In [36]:
autos.head()

date_crawled                                               name  price  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   
2      20160326                         Volkswagen_Golf_1.6_United   8990   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   4350   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   1350   

   ab_test vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004   manually       158  andere   
1  control    limousine               1997  automatic       286     7er   
2     test    limousine               2009   manually       102    golf   
3  control   kleinwagen               2007  automatic        71  fortwo   
4     test        kombi               2003   manually         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot                no   
1       150000                   6    petrol         bmw                no   
2        70000                   7    petrol  volkswagen                no   
3        70000                   6    petrol       smart                no   
4       150000                   7    petrol        ford                no   

   ad_created  postal_code  last_seen  
0    20160326        79588   20160406  
1    20160404        71034   20160406  
2    20160326        35394   20160406  
3    20160312        33729   20160315  
4    20160401        39218   20160401

## Finding the Most Common Brand-Model Combinations

In [37]:
brand = autos['brand']
model = autos['model']
autos['brand_model'] = brand + '/' + model
autos['brand_model'].value_counts().head(5)



volkswagen/golf      3707
bmw/3er              2615
volkswagen/polo      1609
opel/corsa           1592
volkswagen/passat    1349
Name: brand_model, dtype: int64

The most common brand/model combinations are volkswagen/golf and bmw/3er

## Spliting the odometer_km into groups and using aggregation to see if average prices follows any patterns based on the mileage.

We will be dividing odometer_km with interval equal to 30000km


In [38]:
print(autos['odometer_km'].value_counts())
autos.loc[autos["odometer_km"].between(0,30000),"odometer_km_group"] = "A"
autos.loc[autos["odometer_km"].between(30001,60000),"odometer_km_group"] = "B"
autos.loc[autos["odometer_km"].between(60001,90000) ,"odometer_km_group"] = "C"
autos.loc[autos["odometer_km"].between(90001,120000),"odometer_km_group"] = "D"
autos.loc[autos["odometer_km"].between(120001,150000) ,"odometer_km_group"] = "E"
autos.loc[autos["odometer_km"] > 150000  ,"odometer_km_group"] = "F"

150000    30085
125000     4857
100000     2058
90000      1673
80000      1375
70000      1187
60000      1128
50000       993
40000       797
5000        785
30000       760
20000       742
10000       241
Name: odometer_km, dtype: int64


In [39]:
# Aggregation by price based on odometer_km_group
km_group_price_mean = {}

for group in autos["odometer_km_group"].unique():
    selected = autos[autos["odometer_km_group"] == group]
    mean_price = selected["price"].mean()
    km_group_price_mean[group] = int(mean_price)
km_group_price_mean

{'A': 15122, 'B': 13721, 'C': 9563, 'D': 8132, 'E': 4107}

Findings includes:
The lower the odometer_km reading, higher the bidding price and the higher the odometer_km reading, lower the bidding price 


### Analysing how much cheaper are cars with damage than their non-damaged counterparts?

In [40]:
damaged_cars = {}

for i in autos['brand'].unique():
    select_rows = autos[(autos['brand']==i)  &  
                        (autos['unrepaired_damage']=='yes')]
    mean_price_damaged = select_rows['price'].mean()
    damaged_cars[i] = mean_price_damaged
print(damaged_cars)

{'chrysler': 1961.2, 'kia': 2117.9268292682927, 'alfa_romeo': 1889.659090909091, 'honda': 1538.5675675675675, 'mercedes_benz': 3921.8192419825073, 'smart': 1445.0222222222221, 'mitsubishi': 1317.9622641509434, 'mazda': 1418.2439024390244, 'dacia': 4459.625, 'toyota': 3327.0, 'lancia': 3248.0, 'rover': 839.5454545454545, 'subaru': 2346.9, 'fiat': 1146.889705882353, 'jeep': 2833.222222222222, 'saab': 456.125, 'ford': 1375.4855072463768, 'lada': 1400.0, 'volvo': 1800.5652173913043, 'daewoo': 559.8, 'daihatsu': 806.1111111111111, 'land_rover': 5223.75, 'porsche': 13454.545454545454, 'renault': 1145.599290780142, 'chevrolet': 3245.7368421052633, 'volkswagen': 2179.4056603773583, 'trabant': 610.6153846153846, 'audi': 3324.6846846846847, 'nissan': 1947.4086021505377, 'seat': 1741.6279069767443, 'mini': 4595.0, 'citroen': 1944.5441176470588, 'skoda': 3565.268656716418, 'suzuki': 1373.7857142857142, 'bmw': 3512.637717121588, 'peugeot': 1372.3957219251338, 'jaguar': 4444.333333333333, 'sonstige_

In [41]:
nondamaged_cars = {}

for i in autos['brand'].unique():
    select_rows = autos[(autos['brand']==i) &
                        (autos['unrepaired_damage']=='no')]
    mean_price_nondamaged = select_rows['price'].mean()
    nondamaged_cars[i] = mean_price_nondamaged
print(nondamaged_cars)

{'chrysler': 4037.7297297297296, 'kia': 6980.5418326693225, 'alfa_romeo': 4770.735159817352, 'honda': 4978.1461538461535, 'mercedes_benz': 9798.396536796537, 'smart': 3927.5159362549803, 'mitsubishi': 4145.570881226054, 'mazda': 5110.772357723577, 'dacia': 6231.616822429906, 'toyota': 5569.304166666667, 'lancia': 3164.6969696969695, 'rover': 1909.4166666666667, 'subaru': 5455.813559322034, 'fiat': 3446.0693568726356, 'jeep': 12626.0, 'saab': 3895.7796610169494, 'ford': 4660.33211512106, 'lada': 3099.7368421052633, 'volvo': 5845.851851851852, 'daewoo': 1182.4565217391305, 'daihatsu': 1732.658536585366, 'land_rover': 21688.51282051282, 'porsche': 50825.14979757085, 'renault': 3103.2743484224966, 'chevrolet': 7355.712121212121, 'volkswagen': 6469.407758742744, 'trabant': 2897.925925925926, 'audi': 10914.959856396867, 'nissan': 5765.225490196079, 'seat': 5183.380952380952, 'mini': 11157.956639566395, 'citroen': 4289.708423326134, 'skoda': 7053.008210180624, 'suzuki': 4774.107981220657, 'bm

In [42]:
damaged_series = pd.Series(damaged_cars).sort_values(ascending=False)
nondamaged_series = pd.Series(nondamaged_cars).sort_values(ascending=False)

damaged_df = pd.DataFrame(damaged_series, columns=['mean_price_damaged'])
damaged_df['mean_price_nondamaged'] = nondamaged_series
damaged_df['difference'] = damaged_df['mean_price_nondamaged'] - damaged_df['mean_price_damaged']
damaged_df

mean_price_damaged  mean_price_nondamaged    difference
porsche               13454.545455           50825.149798  37370.604343
sonstige_autos         6375.333333           15705.657980   9330.324647
land_rover             5223.750000           21688.512821  16464.762821
mini                   4595.000000           11157.956640   6562.956640
dacia                  4459.625000            6231.616822   1771.991822
jaguar                 4444.333333           14034.701754   9590.368421
mercedes_benz          3921.819242            9798.396537   5876.577295
skoda                  3565.268657            7053.008210   3487.739553
bmw                    3512.637717            9437.709980   5925.072263
toyota                 3327.000000            5569.304167   2242.304167
audi                   3324.684685           10914.959856   7590.275172
lancia                 3248.000000            3164.696970    -83.303030
chevrolet              3245.736842            7355.712121   4109.975279
jeep                   2833.222222           12626.000000   9792.777778
hyundai                2417.631579            6194.592593   3776.961014
subaru                 2346.900000            5455.813559   3108.913559
volkswagen             2179.405660            6469.407759   4290.002098
kia                    2117.926829            6980.541833   4862.615003
chrysler               1961.200000            4037.729730   2076.529730
nissan                 1947.408602            5765.225490   3817.816888
citroen                1944.544118            4289.708423   2345.164306
alfa_romeo             1889.659091            4770.735160   2881.076069
volvo                  1800.565217            5845.851852   4045.286634
seat                   1741.627907            5183.380952   3441.753045
honda                  1538.567568            4978.146154   3439.578586
smart                  1445.022222            3927.515936   2482.493714
mazda                  1418.243902            5110.772358   3692.528455
lada                   1400.000000            3099.736842   1699.736842
ford                   1375.485507            4660.332115   3284.846608
suzuki                 1373.785714            4774.107981   3400.322267
peugeot                1372.395722            3674.019408   2301.623686
opel                   1367.032149            3660.528302   2293.496153
mitsubishi             1317.962264            4145.570881   2827.608617
fiat                   1146.889706            3446.069357   2299.179651
renault                1145.599291            3103.274348   1957.675058
rover                   839.545455            1909.416667   1069.871212
daihatsu                806.111111            1732.658537    926.547425
trabant                 610.615385            2897.925926   2287.310541
daewoo                  559.800000            1182.456522    622.656522
saab                    456.125000            3895.779661   3439.654661

General observation is that non damaged price is higher than their damaged counter part.

## Conclusion

In this project, I applied a variety of pandas methods to explore and understand a dataset on car listings on Ebay.
